feat: enable type-checking with mypy and poe pipeline

nilay-v3rma opened the pull request:
## Towards issue #249 (subtask of #247)

Completed subtasks are ticked below:

- [x] Add `py.typed` file
- [x] Add `mypy` to dev dependencies
- [x] Configure `mypy`
- [x] Add local `poe` pipelines
- [ ] Add workflow for GitHub Actions (pending)

## **Dependency Version Question**
I have set the required `mypy` version in dev dependencies to `>=1.0.0`, while the latest version is `1.15.0`.  
Should I update it to a more recent version, or is `>=1.0.0` fine?

Additionally, I have added the `poe` pipeline for the `prolif` and `tests` directories, as observed in issue #248, is that fine?

## **References for `mypy.ini` configuration:**
1. [Mypy Configuration File](https://mypy.readthedocs.io/en/stable/config_file.html#config-file)
2. [Using Mypy with an Existing Codebase](https://mypy.readthedocs.io/en/stable/existing_code.html#using-mypy-with-an-existing-codebase)

cbouysset commented on pyproject.toml:
```diff
@@ -135,6 +136,10 @@ cmd = "sphinx-build -M html docs/ docs/_build/"
 help = "Runs all of the checks"
 sequence = [{ ref = "style-check" }, { ref = "test" }, { ref = "docs" }]
 
+[tool.poe.tasks.mypy]
```
Just to align with the other `poe` tasks, could you name it after the action (i.e. `type-check`) instead of the tool? This way if in the future we decide to use an alternative to `mypy` contributors won't have to know what's the name of the new tool and whatnot

cbouysset commented on pyproject.toml:
```diff
@@ -135,6 +136,10 @@ cmd = "sphinx-build -M html docs/ docs/_build/"
 
+[tool.poe.tasks.mypy]
+help = "Run mypy to check type hints across the codebase"
+cmd = "mypy prolif tests"
```
We could make the command accept filepaths as a parameter (e.g. to only run the type checking on a specific file) using the syntax for positional args shown [in the poe docs](https://poethepoet.natn.io/guides/args_guide.html#task-argument-options:~:text=type%20boolean.-,multiplebool%20%7C%20int,-If%20the%20multiple).
So that we can run
```
uv run poe type-check
```
to run it on the default paths (I think `.` is enough if we use the `exclude` wisely in the config), or
```
uv run poe type-check prolif/fingerprint.py
```
to run it on specific file(s).

cbouysset reviewed (changes requested):
Thanks @nilay-v3rma for the contribution!

A couple of comments to get you started but this is already looking pretty good.

I think for now we can forget about enabling the command in the CI pipeline as the addition of type hints will come gradually instead of a single PR. Maybe an alternative would be to have a bot comment with the number of typing errors and a `<summary>` element that one can click on to reveal the output of mypy for the files that were modified. Not sure how complicated it would be to add so not a requirement for this PR but would be good to investigate.

nilay-v3rma commented:
Thank you for the detailed review! @cbouysset  
Will work on the suggested tasks asap!

nilay-v3rma commented:
**Summary of the changes made based on review**:

1. Removed `disallow_untyped_calls = True`
2. Removed `disallow_incomplete_defs = True`
3. Removed `check_untyped_defs = True`
4. Changed `python_version` to 3.10
5. Removed `ignore_missing_imports = True` from the general config and enabled it for suggested ones
6. Deleted the `mypy.ini` file and moved config to `pyproject.toml` under `[tool.mypy]`
7. Updated the suggested config settings
8. Made a new "type-check" dependency group and moved `mypy` there
9. Renamed the Poe pipeline to "type-check" based on the action
10. Made the pipeline more dynamic by allowing to accept specific files for checking, keeping `.` as default

cbouy commented on pyproject.toml:
```diff
@@ -68,11 +68,13 @@ dev = [
     { include-group = "tests" },
     { include-group = "docs" },
 ]
+type-check = [
```
could you move this group before the `dev` one, and add `{ include-group = "type-check" },` in `dev`? We need mypy to be installed as part of the dev dependencies

cbouy commented on pyproject.toml:
```diff
@@ -136,9 +138,26 @@ cmd = "sphinx-build -M html docs/ docs/_build/"
 help = "Run mypy to check type hints across the codebase"
-cmd = "mypy prolif tests"
+cmd = "mypy ${FILE_PATHS}"
+args = [{ name = "FILE_PATHS", positional = true, multiple = true, default = "." }]
```
nice!

cbouy commented on pyproject.toml:
```diff
@@ -135,6 +138,27 @@ cmd = "sphinx-build -M html docs/ docs/_build/"
 help = "Runs all of the checks"
 sequence = [{ ref = "style-check" }, { ref = "test" }, { ref = "docs" }]
 
+[tool.mypy]
```
For `exclude` I think we just need to specify the `docs/conf.py` file for now

cbouy commented on uv.lock:
```diff
@@ -1,4 +1,5 @@
 version = 1
+revision = 1
```
Was this added just to resync between what's installed and what's specified in the pyproject dependencies? I don't think we need to keep it

cbouy reviewed (commented):
Nice work, almost there!

nilay-v3rma commented:
I've addressed all the suggested changes. Let me know if there's anything else! :)

cbouysset commented on pyproject.toml:
```diff
@@ -64,13 +64,17 @@ docs = [
     "sphinx-book-theme",
     "prolif[tutorials]",
 ]
+type-check = [
```
Unfortunately I couldn't run the command because of an error in the type-stub files provided by rdkit that was fixed in version `2024.03.2` as seen [here](https://github.com/rdkit/rdkit/issues/7378), so we need to add `rdkit>=2024.03.2` in there. And possibly others like `pandas-stubs` and `types-tqdm`

cbouysset reviewed (changes requested):
One last thing to fix as I wasn't able to run the command because of the rdkit version.

nilay-v3rma commented:
Thanks for pointing that out! I've added the necessary dependencies to the `type-check` group to resolve the error:  

- `rdkit>=2024.03.2` (to include the fixed type-stub files)  
- `pandas-stubs` and `types-tqdm` 

Let me know if anything else needs adjustment! 😊

nilay-v3rma commented:
> Thanks @nilay-v3rma for the contribution!
> 
> A couple of comments to get you started but this is already looking pretty good.
> 
> I think for now we can forget about enabling the command in the CI pipeline as the addition of type hints will come gradually instead of a single PR. Maybe an alternative would be to have a bot comment with the number of typing errors and a `<summary>` element that one can click on to reveal the output of mypy for the files that were modified. Not sure how complicated it would be to add so not a requirement for this PR but would be good to investigate.

I assume if the changes so far are fine, the next task would be figuring out an alternative solution for the CI pipeline, right?

cbouy commented:
Yes, I'll try to run the pipeline when I'm home, and then we can merge this PR since it's already usable by others. Actually, writing these lines I just realized the instructions in the CONTRIBUTING file haven't been updated to reflect your changes, it would be good to have that as well (and mention the possibility to type check specific files, as opposed to the whole repo)

nilay-v3rma commented:
> Checked the latest changes, all good now, we'll just need to:
> 
> * add some additional dependencies for type stubs: `types-setuptools` and `scipy-stubs`,
> * and the CONTRIBUTING.md update under `Getting started > Make changes > 4. Verify that your changes pass tests and other requirements.` you can create a section `d. Running type checks` with the command and optional filepaths positional arg
> 
> And that will be it for this PR!

Thanks for the guidance, I will complete the mentioned tasks asap!

nilay-v3rma commented:
I’ve pushed the updates requested (added the stubs and updated CONTRIBUTING.md). Hope this helps! 😊

nilay-v3rma commented:
> LGTM! Thank you @nilay-v3rma
> 
> For the CI pipeline thing (or should I say the bot commenting on PR with the current progress), could you open an issue specifically for it and whoever wants to handle that can give it a go?

Thank you! 😊 For the bot thing, should I specifically mention Codecov, or should I leave it open-ended for others to choose a different approach?

cbouy commented:
No I haven't really looked into a specific bot, at the end the thing could look like the codecov message that gets posted on PRs, but I'm quite open. Sure I can review it, you can post the draft here (I think you can still comment after I've merged this PR)

nilay-v3rma commented:
> No I haven't really looked into a specific bot, at the end the thing could look like the codecov message that gets posted on PRs, but I'm quite open. Sure I can review it, you can post the draft here (I think you can still comment after I've merged this PR)

Thanks for reviewing in advance! following is the draft,

Implement CI for Type-Check Test Summary Comments on PRs
## Overview
The idea is to implement a CI pipeline that automatically runs type checks on every pull request and posts a comment summarizing the results. The comment should use a `<summary>` tag to collapse detailed error information, providing a concise view with the option to expand for more details.

## Possible Solution
- **CI Integration:**  
  Integrate the type-check command (`uv run poe type-check`) into the CI workflow to run on every commit or PR.
  
- **Commenting:**  
  Develop or configure a mechanism that:
  - Parses the output from the type-check command.
  - Posts a comment on the PR that includes a `<summary>` tag wrapping the detailed type-check results.
  - Updates the comment on subsequent commits to avoid cluttering the PR with multiple comments. (optional)

## Tasks
- [ ] Research existing GitHub Actions or bots that support PR commenting and determine if one can be adapted for our needs.
- [ ] Integrate the chosen bot/tech into our CI pipeline.
- [ ] Test the solution to ensure that it updates the comment appropriately on new commits.

nilay-v3rma commented:
> > No I haven't really looked into a specific bot, at the end the thing could look like the codecov message that gets posted on PRs, but I'm quite open. Sure I can review it, you can post the draft here (I think you can still comment after I've merged this PR)
> 
> Thanks for reviewing in advance! following is the draft,
> 
> Title: Implement CI for Type-Check Test Summary Comments on PRs
> 
> ## Overview
> The idea is to implement a CI pipeline that automatically runs type checks on every pull request and posts a comment summarizing the results. The comment should use a `<summary>` tag to collapse detailed error information, providing a concise view with the option to expand for more details.
> 
> ## Possible Solution
> * **CI Integration:**
>   Integrate the type-check command (`uv run poe type-check`) into the CI workflow to run on every commit or PR.
> * **Commenting:**
>   Develop or configure a mechanism that:
>   
>   * Parses the output from the type-check command.
>   * Posts a comment on the PR that includes a `<summary>` tag wrapping the detailed type-check results.
>   * Updates the comment on subsequent commits to avoid cluttering the PR with multiple comments. (optional)
> 
> ## Tasks
> * [ ]  Research existing GitHub Actions or bots that support PR commenting and determine if one can be adapted for our needs.
> * [ ]  Integrate the chosen bot/tech into our CI pipeline.
> * [ ]  Test the solution to ensure that it updates the comment appropriately on new commits.

@cbouy, tagging you in case you didn't get notified about the message. Sincere apologies if this was unnecessary.

cbouysset commented:
That's really good @nilay-v3rma, I'd just keep the title of the issue a bit shorter e.g. `Implement PR bot for type-checking`, all the rest is great! Please ping both of my accounts in there once you open it.
Thanks a lot 🙌

nilay-v3rma commented:
> That's really good @nilay-v3rma, I'd just keep the title of the issue a bit shorter e.g. `Implement PR bot for type-checking`, all the rest is great! Please ping both of my accounts in there once you open it. Thanks a lot 🙌

Done! :)
